agent/src/main/java/io/pyroscope/javaagent/impl/ContinuousProfilingScheduler.java

```diff
@@ -38,14 +39,76 @@ public ContinuousProfilingScheduler(Config config, Exporter exporter, Logger log
 
     @Override
     public void start(Profiler profiler) {
-        Duration firstProfilingDuration;
+        this.logger.log(Logger.Level.DEBUG, "ContinuousProfilingScheduler#start");
+        synchronized (lock) {
+            if (started) {
+                throw new IllegalStateException("already started");
+            }
+            Duration firstProfilingDuration;
+            try {
+                firstProfilingDuration = startFirst(profiler);
+            } catch (Throwable throwable) {
+                stopSchedulerLocked();
+                throw new IllegalStateException(throwable);
+            }
+            this.profiler = profiler;
+            this.executor = Executors.newSingleThreadScheduledExecutor(THREAD_FACTORY);
+            this.job = executor.scheduleAtFixedRate(this::schedulerTick,
+                firstProfilingDuration.toMillis(), config.uploadInterval.toMillis(), TimeUnit.MILLISECONDS);
+            this.started = true;
+        }
+    }
+
+    @Override
+    public void stop() {
+        ScheduledExecutorService svc = this.executor;
+        synchronized (lock) {
+            stopSchedulerLocked();
+            this.executor = null;
+        }
+        if (svc != null) {
+            awaitTermination(svc);
+        }
+        this.logger.log(Logger.Level.DEBUG, "ContinuousProfilingScheduler stopped");
+    }
+
+    private static void awaitTermination(ScheduledExecutorService svc) {
+        try {
+            boolean terminated = svc.awaitTermination(10, TimeUnit.SECONDS);
+            if (!terminated) {
+                throw new IllegalStateException("failed to terminate scheduler's executor");
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("failed to terminate scheduler's executor", e);
+        }
+    }
+
+    private void stopSchedulerLocked() {
+        if (!this.started) {
+            return;
+        }
+        this.logger.log(Logger.Level.DEBUG, "ContinuousProfilingScheduler stopping");
         try {
-            firstProfilingDuration = startFirst(profiler);
+            this.profiler.stop();
         } catch (Throwable throwable) {
-            stop();
-            throw new IllegalStateException(throwable);
+            logger.log(Logger.Level.ERROR, "Error stopping profiler %s", throwable);
```

Comment:
This used to throw but now it doesn't, is this intentional?

Reply:
not sure I understand, there were no stop functionality before, so it did not throw.

However I see the inconsistency, start throws IllegalStateException on profiler.start fail, and stop does not. I will try to make it consistent.